src/lat_grid.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man
.TH LAT_GRID  "1" " " "lat_grid http://www.antiprism.com" "User Commands"
.SH NAME
lat_grid - lattices and grids with integer coordinates
.SH SYNOPSIS
.B lat_grid
[\fI\,options\/\fR] [\fI\,lat_type \/\fR[\fI\,outer_width \/\fR[\fI\,inner_width\/\fR]]]
.SH DESCRIPTION
Make a lattice or grid with integer coordinates. Lattice types (default: sc)
are followed by name and then and valid strut arguments (squares of length,
first is square of the radius of balls packed at the vertex positions)
.TP
sc
\- Simple Cubic                         (1, 2, 3)
.TP
fcc
\- Face Centred Cubic                   (2, 4, 6, 12)
.TP
bcc
\- Body Centred Cubic                   (3, 4, 8)
.TP
hcp
\- Hexagonal Close Packing              (18)
.TP
rh_dodec
\- Rhombic Dodecahedra                  (3, 8)\e
.TP
cubo_oct
\- Cuboctahedron / Octahedron           (2)
.TP
tr_oct
\- Truncated Octahedron                 (2)
.TP
tr_tet_tet
\- Truncated Tetrahedron / Tetrahedron  (2)
.TP
tr_tet_tr_oct_cubo \- Truncated Tetrahedron /
Truncated Octahedron / Cuboctahedron (4)
.TP
diamond
\- Diamond                              (3)
.TP
hcp_diamond
\- HCP Diamond                          (27)
.TP
k_4
\- K_4 Crystal                          (2)
.PP
Inner and outer widths are the sizes of the inner and outer containers.
For cubes, these are the length of a side, for spheres these are the
squares of the radii.
.PP
Options
.HP
\fB\-h\fR,\-\-help this help message (run 'off_util \fB\-H\fR help' for general help)
.HP
\fB\-\-version\fR version information
.HP
\fB\-C\fR <cent> centre of lattice, in form \e"x_val,y_val,z_val\e"
.HP
\fB\-c\fR <type> container, c \- cube (default), s \- sphere
.HP
\fB\-s\fR <len2> create struts, the value is the square of the strut length
.HP
\fB\-o\fR <file> write output to file (default: write to standard output)
.SH "SEE ALSO"
The full documentation for
.B lat_grid
is maintained as a Texinfo manual.  If the
.B info
and
.B lat_grid
programs are properly installed at your site, the command
.IP
.B info lat_grid
.PP
should give you access to the complete manual.


NEWS
NEWS for Antiprism: http://www.antiprism.com

Version 0.28
xx-xx-20 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            poly_form: rename from minmax, rename options, add -y
              for symmetry constraint, add -a e equal-edge algorithm
              and -a p planarisation
         Changes
            to_nfold: add -x, -p to help with crossed models
            minmax: rename to poly_form
            repel: add -z for status reporting
            rotegrity: remove -T (old twist program output)
         Fixes
            minor bug fixes and improvements
            code review and fixes for potential issues

Version 0.27
01-10-20 Adrian Rossiter <adrian@antiprism.com>
         Changes
            off_color: new -e jJ to colour by edge length
            zono: -P add spirallohedra, new -T translation product
            conway: g and s take a numeric parameter
         Fixes
            off2vrml: avoid duplicate PROTO statements
            minor bug fixes and improvements

Version 0.26
05-04-19 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            rotegrity: make rotegrity and nexorade models
            mmop_origami: multimodular polyhedron origami models
         Changes
            antiview: new keys 'u/U' change label size, 'M' planar
              face labels
            wythoff: new option -C for colouring by base elements
            conway: new options -C colouring method, -R colour seed
            geodesic, geo_m_n: change direction of pattern turn,
              so (m, n) is like previous (n, m)
            bravais: added inverse of f and i centering types
            n_icons: swap -J and -L option letters
            radial: renamed to off_color_radial
            twist: removed (use rotegrity -T -f)
            pol_recip: removed -r X (use wythoff -c d)
            sph_saff: removed (use antiprism_python)
            off2m,m2off: removed (Java applets no longer popular)
         Fixes
            minor bug fixes and improvements

Version 0.25.1
08-12-17 Adrian Rossiter <adrian@antiprism.com>
         Fixes
            build: issues affecting OSX, install all headers

Version 0.25
23-11-17 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            stellate: polyhedron stellations (Roger Kaufman)
            miller: Miller's stellations of the icosahedron (Roger Kaufman)
            wythoff: Wythoff-style tilings on various surfaces
            kaleido: Legacy program ported to Antiprism (Roger Kaufman)
            radial: Color by symmetry radiating from axes (Roger Kaufman)
         Changes
            poly_weave: moved to unsupported programs
            off2obj: added optional mtl file output
            canonical: works on more models and code reorganization
              added new planarization and display options
            conway: changed to use wythoff engine while keeping legacy
              code by George Hart. Enhancements and additional operators
              to original conway operators
            waterman: option to fill interior points
            off_util: removed -E
         Fixes
            pol_recip: dual vertices for very small faces
            poly_weave: input can be non-closed surface
            off_util: -u input can be non-closed surface, -T works with
              more models
            antiview: default handling of large models

Version 0.24.1
18-11-16 Adrian Rossiter <adrian@antiprism.com>
         Fixes
            symmetro: docs

Version 0.24
08-10-16 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            to_nfold: change the rotational symmetry of a polyhedron
            symmetro: polyhedra with regular polygons at symmetry
              axes (Roger Kaufman)
            off2obj/obj2off: convert between OFF to Wavefront OBJ
              format (Roger Kaufman)
            off2dae: convert OFF to Collada (DAE) format
         Changes
            off_util: process options in command line order,
              incorporate -b into -M, expand -D and add -K to keep
              elements (Roger Kaufman), expand documentation.
            minmax: new option -a u, make faces regular. Improve
              reporting.
            leonardo: orient models, new options -c centre, -p panel faces,
              -k don't orient
            zono: -P add nonconvex models, new option -S zonish polyhedra
            off_normals: options -t and -m removed
            off_color: new edge direction colourings -e ldD
            antiview: 'O' to show orientation, 't' includes 50%
              transparency
            build: native Debian/Ubuntu package build, support FLTK GLUT
         Fixes
            off_util -u: natural unfolding for positive orientation
            off_align: fix crashes
            off2pov: fix compatibility with POV-Ray 3.7
            symmetry: fix missing subsyms
            minor bug fixes

Version 0.23
11-03-13 Adrian Rossiter <adrian@antiprism.com>
         New Programs
            iso_kite: isohedral kite-faced polyhedra
         Changes
            Floating point arguments can be mathematical expressions
              (sqrt prefix changed to rt, e.g. rt5)
            off_util: merge preserves original element order
            off_query: report element colours
            off_color: GgCc colourings now use z-axis for height.
            polygon: new crown polyhdron types
            planar: -W takes a colouring argumant
            Symmetry: symmetry alignment should not realign an
              aligned polyhedron, element orbits ordered by lowest
              index number
            Resource models: Johnson models are symmetry aligned
              (append _raw for old alignment).
            Build: install directories do not include version number
         Fixes
            leonardo: -o outputs to file 
            Build: support 'configure --docdir' option, fix errors
              and warnings when building with clang

Version 0.22

src/lat_grid.cc
/*
   Copyright (c) 2003-2016, Adrian Rossiter

   Antiprism - http://www.antiprism.com

   Permission is hereby granted, free of charge, to any person obtaining a
   copy of this software and associated documentation files (the "Software"),
   to deal in the Software without restriction, including without limitation
   the rights to use, copy, modify, merge, publish, distribute, sublicense,
   and/or sell copies of the Software, and to permit persons to whom the
   Software is furnished to do so, subject to the following conditions:

      The above copyright notice and this permission notice shall be included
      in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
  IN THE SOFTWARE.
*/

/*
   Name: lat_grid.cc
   Description: program to make lattices and grids with intger coordinates
   Project: Antiprism - http://www.antiprism.com
*/

#include "../base/antiprism.h"
#include "lattice_grid.h"

#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

using std::string;
using std::vector;

using namespace anti;

class lg_opts : public ProgramOpts {
public:
  double o_width;
  double i_width;
  Vec3d centre;
  int strut_len2;
  COORD_TEST_F coord_test;
  char container;

  string ofile;

  lg_opts()
      : ProgramOpts("lat_grid"), o_width(6), i_width(-1), strut_len2(0),
        coord_test(sc_test), container('c')
  {
  }

  void process_command_line(int argc, char **argv);
  void usage();
};

void lg_opts::usage()
{
  fprintf(stdout, R"(
Usage: %s [options] [lat_type [outer_width [inner_width]]]

Make a lattice or grid with integer coordinates. Lattice types (default: sc)
are followed by name and then and valid strut arguments (squares of length,
first is square of the radius of balls packed at the vertex positions)
  sc                 - Simple Cubic                         (1, 2, 3)
  fcc                - Face Centred Cubic                   (2, 4, 6, 12)
  bcc                - Body Centred Cubic                   (3, 4, 8)
  hcp                - Hexagonal Close Packing              (18)
  rh_dodec           - Rhombic Dodecahedra                  (3, 8)\
  cubo_oct           - Cuboctahedron / Octahedron           (2)
  tr_oct             - Truncated Octahedron                 (2)
  tr_tet_tet         - Truncated Tetrahedron / Tetrahedron  (2)
  tr_tet_tr_oct_cubo - Truncated Tetrahedron / 
                       Truncated Octahedron / Cuboctahedron (4)
  diamond            - Diamond                              (3)
  hcp_diamond        - HCP Diamond                          (27)
  k_4                - K_4 Crystal                          (2)

Inner and outer widths are the sizes of the inner and outer containers.
For cubes, these are the length of a side, for spheres these are the
squares of the radii.

Options
%s
  -C <cent> centre of lattice, in form \"x_val,y_val,z_val\"
  -c <type> container, c - cube (default), s - sphere
  -s <len2> create struts, the value is the square of the strut length
  -o <file> write output to file (default: write to standard output)

)",
          prog_name(), help_ver_text);
}

void lg_opts::process_command_line(int argc, char **argv)
{
  opterr = 0;
  int c;

  handle_long_opts(argc, argv);

  while ((c = getopt(argc, argv, ":hc:s:C:o:")) != -1) {
    if (common_opts(c, optopt))
      continue;

    switch (c) {
    case 's':
      print_status_or_exit(read_int(optarg, &strut_len2), c);
      if (strut_len2 < 0)
        error("square of strut length cannot be negative", "s");
      break;

    case 'C':
      print_status_or_exit(centre.read(optarg), c);
      break;

    case 'c':
      if (strlen(optarg) > 1 || !strchr("cs", *optarg))
        error("method is '" + string(optarg) + "' must be c or s");
      container = *optarg;
      break;

    case 'o':
      ofile = optarg;
      break;

    default:
      error("unknown command line error");
    }
  }

  if (argc - optind > 3) {
    error("too many arguments");
    exit(1);
  }

  if (optind < argc) {
    string lt = argv[optind];
    if (lt == "sc")
      coord_test = sc_test;
    else if (lt == "fcc")
      coord_test = fcc_test;
    else if (lt == "bcc")
      coord_test = bcc_test;
    else if (lt == "hcp")
      coord_test = hcp_test;
    else if (lt == "rh_dodec")
      coord_test = rh_dodec_test;
    else if (lt == "cubo_oct")
      coord_test = cubo_oct_test;
    else if (lt == "tr_oct")
      coord_test = tr_oct_test;
    else if (lt == "tr_tet_tet")
      coord_test = tr_tet_tet_test;
    else if (lt == "tr_tet_tr_oct_cubo")
      coord_test = tr_tet_tr_oct_cubo_test;
    else if (lt == "diamond")
      coord_test = diamond_test;
    else if (lt == "hcp_diamond")
      coord_test = hcp_diamond_test;
    else if (lt == "k_4")
      coord_test = k_4_test;
    else
      error("unknown lattice type '" + lt + "' (try " + prog_name() + " -h)",
            "lat_type");
  }

  optind++;
  if (optind < argc) {
    print_status_or_exit(read_double(argv[optind], &o_width), "outer_width");
    if (o_width < 0)
      error("cannot be negative", "outer_width");
  }

  optind++;
  if (optind < argc) {
    print_status_or_exit(read_double(argv[optind], &i_width), "inner_width");
    if (i_width > o_width)
      error("cannot be greater than outer container width", "inner_width");
  }
}

int main(int argc, char **argv)
{
  lg_opts opts;
  opts.process_command_line(argc, argv);

  int_lat_grid *lat;
  if (opts.container == 'c')
    lat = new int_lat_grid();
  else
    lat = new sph_lat_grid();

  lat->set_o_width(opts.o_width);
  lat->set_i_width(opts.i_width);
  lat->set_centre(opts.centre);
  lat->set_coord_test(opts.coord_test);

  Geometry geom;
  lat->make_lattice(geom);
  delete lat;

  if (opts.strut_len2 > 0)
    add_struts(geom, opts.strut_len2);

  opts.write_or_error(geom, opts.ofile);

  return 0;
}
